Add a generic type for sharing some CSS types

Wafflespeanut opened the pull request:
<!-- Please describe your changes on the following line: -->

This would be useful for types such as `T(pub Option<U>)` or any type that's a mixture of two types (like `LengthOrFoo` for example). Well, it's a bit ugly, especially because we have to address the types as `Either::First` and `Either::Second` everywhere, but I don't have a brighter idea :smile: 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [x] These changes do not require tests because it's a refactor

r? @SimonSapin (cc @Manishearth @emilio)

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14049)
<!-- Reviewable:end -->

Commit:
Add an Either<A, B> type for sharing specified and computed CSS types

Commit:
Add unit structs for representing keywords: auto, none, normal

Commit:
Make use of Either<A, B> for LengthOrNone

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/longhand/effects.mako.rs, components/style/values/mod.rs, components/style/values/specified/mod.rs, components/style/values/computed/mod.rs, components/style/properties/helpers/animated_properties.mako.rs, components/style/properties/properties.mako.rs
 * @emilio: components/style/properties/longhand/effects.mako.rs, components/style/values/mod.rs, components/style/values/specified/mod.rs, components/style/values/computed/mod.rs, components/style/properties/helpers/animated_properties.mako.rs, components/style/properties/properties.mako.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

Manishearth commented on components/style/properties/helpers/animated_properties.mako.rs:
```diff
@@ -681,8 +682,8 @@ impl Interpolate for BoxShadow {
 impl Interpolate for LengthOrNone {
     fn interpolate(&self, other: &Self, progress: f64) -> Result<Self, ()> {
         match (*self, *other) {
-            (LengthOrNone::Length(ref len), LengthOrNone::Length(ref other)) =>
```
I think we should typedef/reexport the most common types.

Wafflespeanut commented:
Oops. I opened the PR once I was able to build `geckolib`. Still needs some work to build servo.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14077) made this pull request unmergeable. Please resolve the merge conflicts.

Wafflespeanut commented:
This is ready for review!

SimonSapin commented:
Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 10 of 10 files at r4, 3 of 7 files at r5.
Review status: 9 of 13 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[components/layout/display_list_builder.rs, line 1584 at r4](https://reviewable.io:443/reviews/servo/servo/14049#-KW3hjajKk57ya3Ryd_B:-KW3hjakdqtlJgfjsYOi:b-nndtmt) ([raw file](https://github.com/servo/servo/blob/50b790633377a2b074fdc6bf5fb1e09709151d86/components/layout/display_list_builder.rs#L1584)):*
> ```Rust
>                 pre_transform.pre_mul(&perspective_matrix).pre_mul(&post_transform)
>             }
>             Either::Second(_none) => {
> ```

Instead of a `_none` ignored binding, this could be the `values::None_` unit struct used as a pattern, to make it clear there is no data we’re ignoring.

---

*[components/layout/display_list_builder.rs, line 1564 at r5](https://reviewable.io:443/reviews/servo/servo/14049#-KW3gEVthT7mj9fYhKH7:-KW3gEVxJTqKN-bJBBDR:ba2ocz0) ([raw file](https://github.com/servo/servo/blob/9e690053933dbec7bb3d7e65f5fc44fb3cedf550/components/layout/display_list_builder.rs#L1564)):*
> ```Rust
> 
>         let transform = self.transform_matrix(&border_box);
>         let perspective = match self.style().get_effects().perspective.as_length() {
> ```

I’m not super sure this is a good change: previously you’d know for sure that `….perspective` has two variants and that we’re exhaustively matching all of them. Now it’s not clear by reading this code (without external knowledge of the `perspective` CSS property, which might be extended in a future CSS spec) that the "not a length" case can only be "the 'none' keyword".

Same comment for every use of `.as_length()`, pretty much.

---

*[components/style/values/mod.rs, line 46 at r2](https://reviewable.io:443/reviews/servo/servo/14049#-KW3W8G7_RzQlFdL0Suc:-KW3W8G87Is2QEqcnTmp:b-ql8el9) ([raw file](https://github.com/servo/servo/blob/cf0d42194539429f6666cb373318c346c1fefc9d/components/style/values/mod.rs#L46)):*
> ```Rust
> }
> 
> macro_rules! impl_specified_for_computed {
> ```

This should probably be a generic impl:

```rust
impl<A: ToComputedValue, B: ToComputedValue> ToComputedValue for Either<A, B> {
    type ComputedValue = Either<A::ComputedValue, B::ComputedValue>;
    fn to_computed_value(&self, context: &Context) -> Self::ComputedValue { … }
    fn from_computed_value(computed: &Self::ComputedValue) -> Self { … }
}

---

*[components/style/values/mod.rs, line 88 at r2](https://reviewable.io:443/reviews/servo/servo/14049#-KW3WqGRznUdeBe0FQte:-KW3WqGRznUdeBe0FQtf:b-5ravdi) ([raw file](https://github.com/servo/servo/blob/cf0d42194539429f6666cb373318c346c1fefc9d/components/style/values/mod.rs#L88)):*
> ```Rust
> }
> 
> pub trait CssType: Debug + FromCss + ToCss {}
> ```

I think this does not need to exist. See comments below.

---

*[components/style/values/mod.rs, line 94 at r2](https://reviewable.io:443/reviews/servo/servo/14049#-KW3WbOr9HspxaMc7Z0S:-KW3WbOr9HspxaMc7Z0T:b-koktve) ([raw file](https://github.com/servo/servo/blob/cf0d42194539429f6666cb373318c346c1fefc9d/components/style/values/mod.rs#L94)):*
> ```Rust
> #[derive(Clone, PartialEq, Copy)]
> #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
> pub enum Either<A: CssType, B: CssType> {
> ```

I think these bounds on the type definition are not needed. On the contrary, they are what forces you to write dummy impls like `impl FromCss for Au`.

---

*[components/style/values/mod.rs, line 99 at r2](https://reviewable.io:443/reviews/servo/servo/14049#-KW3X1E27lldk_n_72sS:-KW3X1E27lldk_n_72sT:bbl5amg) ([raw file](https://github.com/servo/servo/blob/cf0d42194539429f6666cb373318c346c1fefc9d/components/style/values/mod.rs#L99)):*
> ```Rust
> }
> 
> impl<A: CssType, B: CssType> Debug for Either<A, B> {
> ```

With the bounds removed from the type definition, these only need to be `impl<A: Debug, B: Debug>`.

Same comment for `ToCss` and `parse` below.

---

*[components/style/values/mod.rs, line 103 at r2](https://reviewable.io:443/reviews/servo/servo/14049#-KW3XCiCy8ojX0ja8v74:-KW3XCiCy8ojX0ja8v75:b-guwus3) ([raw file](https://github.com/servo/servo/blob/cf0d42194539429f6666cb373318c346c1fefc9d/components/style/values/mod.rs#L103)):*
> ```Rust
>         match *self {
>             Either::First(ref v) => write!(f, "{:?}", v),
>             Either::Second(ref v) => write!(f, "{:?}", v),
> ```

Since you already have a `fmt::Formatter`, calling `v.fmt(f)` is probably more efficient than going through the whole `write!` machinery again.

---

*[components/style/values/mod.rs, line 121 at r2](https://reviewable.io:443/reviews/servo/servo/14049#-KW3XSxOrhGNKb6ATLRH:-KW3XSxOrhGNKb6ATLRI:bi1ldg1) ([raw file](https://github.com/servo/servo/blob/cf0d42194539429f6666cb373318c346c1fefc9d/components/style/values/mod.rs#L121)):*
> ```Rust
>         if let Ok(v) = input.try(|i| A::from_css(i)) {
>             Ok(Either::First(v))
>         } else if let Ok(v) = B::from_css(input) {
> ```

Nit: this is slightly simpler:

```rust
} else {
    B::from_css(input).map(Either::Second)
}

---

*[components/style/values/mod.rs, line 170 at r5](https://reviewable.io:443/reviews/servo/servo/14049#-KW3foVdSQfvjaU2SZbd:-KW3foVdSQfvjaU2SZbe:b-wc7e64) ([raw file](https://github.com/servo/servo/blob/9e690053933dbec7bb3d7e65f5fc44fb3cedf550/components/style/values/mod.rs#L170)):*
> ```Rust
> }
> 
> macro_rules! impl_either_type_viewport_percent {
> ```

I think that, like every trait impl for Either, this can be a generic impl: `impl<A: HasViewportPercentage, B: HasViewPortPercentage> HasViewportPercentage for Either<A, B>`

---

*[components/style/values/specified/length.rs, line 164 at r4](https://reviewable.io:443/reviews/servo/servo/14049#-KW3_fbGrUh_1JLoN9vZ:-KW3_fbHXuk8h9Sdap3j:b-gx0057) ([raw file](https://github.com/servo/servo/blob/50b790633377a2b074fdc6bf5fb1e09709151d86/components/style/values/specified/length.rs#L164)):*
> ```Rust
> 
> impl FromCss for Length {
>     fn from_css(input: &mut Parser) -> Result<Self, ()> {
> ```

This should probably be called `parse` and replace the inherent `parse` method, not just call it.

---

*[components/style_traits/values.rs, line 66 at r1](https://reviewable.io:443/reviews/servo/servo/14049#-KW3Td_1Gs295qrzlnYP:-KW3Td_2dAb9NeXJX69B:bo22rnd) ([raw file](https://github.com/servo/servo/blob/0796b8d01735b42effbe88fcef177dee3444899e/components/style_traits/values.rs#L66)):*
> ```Rust
> 
> /// We don't parse any CSS values to `Au`, since it's only used for representing
> /// computed lengths. We implement a garbage `FromCss` here only for the sake of
> ```

Why do computed lengths need FromCss at all?

If this impl *is* needed but never used, please make it panic with `unimplemented!()` rather than silently return a dummy value, so that we realize if it’s used accidentally.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14049)*
<!-- Sent from Reviewable.io -->

Wafflespeanut commented:
Addressed all the comments. `FromCss`, `CssType`, enum bounds, macros have all been burned! It's quite clean now :smile: 

---

Review status: 1 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[components/style_traits/values.rs, line 66 at r1](https://reviewable.io:443/reviews/servo/servo/14049#-KW3Td_1Gs295qrzlnYP:-KW6yyLE7YPE8ImVETHb:bb2gf3l) ([raw file](https://github.com/servo/servo/blob/0796b8d01735b42effbe88fcef177dee3444899e/components/style_traits/values.rs#L66)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary>
> Why do computed lengths need FromCss at all?
> 
> If this impl *is* needed but never used, please make it panic with `unimplemented!()` rather than silently return a dummy value, so that we realize if it’s used accidentally.
</details>
Nah, it's not necessary. As you said, it was because of the bound enforced in the enum :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14049#-:-KW6z4-KQaCLhqer3llG:b-hinutx)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
@bors-servo r+

Good work, thanks!

---

Reviewed 1 of 1 files at r7, 10 of 10 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14049#-:-KW8RGgc35szuCsiFiL-:b-gv8awc)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit c4fc49c has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin c4fc49c55908f62b3a63ab72da492db9017d1aff -->

bors-servo commented:
:hourglass: Testing commit c4fc49c with merge 35f328d...

Wafflespeanut commented:
Whoo! Thanks for the review! :)

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3997), [arm64](http://build.servo.org/builders/arm64/builds/4003), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4096), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/739), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/738), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4081), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4077), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/761), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/774), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2009)
